### ISSUE TITLE
Windows dialogs displays shaded lines

### ISSUE THREAD

**Issue by qianniancn** [fyne-io/fyne#2471]
### Screenshots:

![20210914210008](https://user-images.githubusercontent.com/5478237/133265377-0c51250e-b63f-4f4c-93b9-bb405a18e7f0.png)
![20210914211958](https://user-images.githubusercontent.com/5478237/133265385-f96fe51b-7981-4619-b209-56cddb55bed5.png)
![20210914212021](https://user-images.githubusercontent.com/5478237/133265389-82da1324-1c3b-4404-bca7-d898e28195b1.png)

### Example code:
fyne demo

### Device (please complete the following information):
 - **OS:** windows10
 - **Version:** 10
 - **Go version:** 1.17
 - **Fyne version:** 2.0.4

**Comment by andydotxyz** [comment]
This will relate to output details. Can you please confirm the type of output device, resolution and what scale your Windows desktop is set to?

**Comment by qianniancn** [comment]
@andydotxyz
scaling:125%
desktop resolution:1920*1080

**Comment by qianniancn** [comment]
![20210914231306](https://user-images.githubusercontent.com/5478237/133284961-2651f03c-9292-4eac-9f5c-74894d977eb0.png)

**Comment by andydotxyz** [comment]
Thanks, that helps.
I am able to replicate this on other OS using FYNE_SCALE=1.25. This is subtle so I see why it slipped by. Probably to do with our shift to floating point sizes in v2.0.

**Comment by andydotxyz** [comment]
I had a little look and it seems to be deep in the rounding code for render.
The `roundToPixel` seems to introduce it for non-integer scale/texturescale combinations.
However I cannot see a good solution as the code is correct.

If you reduce the precision in that method the issue amplifies, so it may be that the error is elsewhere but that this rounding code is reducing it to a single on-screen pixel difference.

**Comment by toaster** [comment]
Yeah, that’s a general issue when transforming floating point coordinates and sizes to integer space.
The issue itself is caused by the shadowed area having a fractional width (or height). This than leads to an off-by-one when computing the right/bottom shadow position, sometimes.

**Comment by ranjith-bhide** [comment]
Hi,

I was trying to debug this issue with a simple fyne Application and here are my findings:

My sample Code:
```go
func main() {

	appInstance := app.NewWithID("appInstance")
	win := appInstance.NewWindow("window")

	simple_bg := canvas.NewRectangle(color.RGBA{50, 100, 150, 255})
	content := container.NewBorder(
		nil,
		widget.NewButton("Press Me !!!", func() {
			dialog.ShowError(errors.New("sample error"), win)
		}),
		nil,
		nil,
		simple_bg,
	)
	win.SetContent(content)
	win.Resize(fyne.NewSize(800, 600))
	win.ShowAndRun()
}
```
And was able to get this Issue even without setting the FYNE_SCALE env variable.
![issue_screenshot](https://user-images.githubusercontent.com/30768539/161378758-60cc215c-4142-4ea6-9285-c562093eb888.png)

While Debugging I found that in file draw.go function roundToPixelCoords() seems to be meddling with the Position and Size Values of the Shadow Elements

**Before roundToPixelCoords()**

Component | X | Y | Width | Height | EndX | EndY
-- | -- | -- | -- | -- | -- | --
Top-Left | 301.8258 | 199.0611 | 24 | 24 | 325.8258 | 223.0611
Top | 325.8258 | 199.0611 | 148.25 | 24 | 474.0758 | 223.0611
Top-Right | 474.0758 | 199.0611 | 24 | 24 | 498.0758 | 223.0611
Right | 474.0758 | 223.0611 | 24 | 154.01562 | 498.0758 | 377.07672
Bottom-Right | 474.0758 | 377.07672 | 24 | 24 | 498.0758 | 401.07672
Bottom | 325.8258 | 377.07672 | 148.25 | 24 | 474.0758 | 401.07672
Bottom-Left | 301.8258 | 377.07672 | 24 | 24 | 325.8258 | 401.07672
Left | 301.8258 | 223.0611 | 24 | 154.01562 | 325.8258 | 377.07672
Dialog box | 325.8258 | 223.0611 | 148.25 | 154.01562 | 474.0758 | 377.07672


We can see that the EndX EndYs of the each element exactly matches with the Next Element's StartX, StartY

However **After roundToPixelCoords()** (PixScale = 1.18)

Component | X | Y | Width | Height | EndX | EndY
-- | -- | -- | -- | -- | -- | --
Top-Left | 302.18506 | 198.91733 | 23.700788 | 23.700788 | 325.885848 | 222.618118
Top | 325.88583 | 198.91733 | 148.12993 | 23.700788 | 474.01576 | 222.618118
Top-Right | 474.01578 | 198.91733 | 23.700788 | 23.700788 | 497.716568 | 222.618118
Right | 474.01578 | 223.46458 | 23.700788 | 154.05513 | 497.716568 | 377.51971
Bottom-Right | 474.01578 | 376.67325 | 23.700788 | 23.700788 | 497.716568 | 400.374038
Bottom | 325.88583 | 376.67325 | 148.12993 | 23.700788 | 474.01576 | 400.374038
Bottom-Left | 302.18506 | 376.67325 | 23.700788 | 23.700788 | 325.885848 | 400.374038
Left | 302.18506 | 223.46458 | 23.700788 | 154.05513 | 325.885848 | 377.51971
Dialog box | 325.88583 | 223.46458 | 148.12993 | 154.05513 | 474.01576 | 377.51971

Here we can see that the Top 3 elements's EndY is around 0.8 units away from the subsequent Center Elements' StartY
and there is about 0.8 units of overlap between EndY of Center Elements and StartY of Bottom Elements.
This explains why there is a sort of White Line artifact near the top and Black Line artifact near the bottom.
The Differenct/Overlap between the EndX and StartX values are very small thus the Vertical Line Artifacts are not visible.

Here is the Proposed Fix:
```diff
diff --git a/internal/painter/gl/draw.go b/internal/painter/gl/draw.go
index 6039cad0..425fcbbc 100644
--- a/internal/painter/gl/draw.go
+++ b/internal/painter/gl/draw.go
@@ -224,10 +224,13 @@ func roundToPixel(v float32, pixScale float32) float32 {
 }
 
 func roundToPixelCoords(size fyne.Size, pos fyne.Position, pixScale float32) (fyne.Size, fyne.Position) {
-       size.Width = roundToPixel(size.Width, pixScale)
-       size.Height = roundToPixel(size.Height, pixScale)
+       end := pos.Add(size)
+       end.X = roundToPixel(end.X, pixScale)
+       end.Y = roundToPixel(end.Y, pixScale)
        pos.X = roundToPixel(pos.X, pixScale)
        pos.Y = roundToPixel(pos.Y, pixScale)
+       size.Width = end.X - pos.X
+       size.Height = end.Y - pos.Y
 
        return size, pos
 }
```
The rationale behind the above fix is as follows:
I suspect this is happening in roundToPixelCoords() function due to rounding of the Pos and Size value separately, where individual values might be floored but the Sum's round off would be expected to be ceiled. 
Eg: Pos=5.4 and Len=6.3 rounded off individually will give Start = 5 and End =5+6 = 11
However its more appropriate to compute End as Round(Pos + Len) = round(11.7) = 12 and then compute the Len as 12-5 = 7
And because both Start and End are rounded off their difference (Len) will be by default rounded off as well.

I am able to Verify this fix to be working and the artifacts are no longer seen with FYNE_SCALE=1.0 1.25, 1.3 and 2

**Comment by andydotxyz** [comment]
Thanks @ranjith-bhide this is an excellent analysis. Would you be interested in opening a PR for this fix so you can get the credit? :)

**Comment by ranjith-bhide** [comment]
:) Thanks @andydotxyz
I've raised a PR with the required change, and I think the Workflow execution needs approval from a maintainer.
I am a bit new to contributing to OSS, So Please let me know if any Action Item is pending on me for this.

**Comment by Jacalz** [comment]
I approved the action for you. Thanks for opening the PR :)

**Comment by andydotxyz** [comment]
Thanks so much, merged onto develop for v2.2.0
